hack/update-client-gen.sh

```diff
@@ -6,36 +6,21 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-REPO_ROOT=$(dirname "${BASH_SOURCE}")/..
-BINDIR=${REPO_ROOT}/bin
-HACKDIR=${HACKDIR:-hack}
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
```

Comment:
❓ I can't quite figure out what this is doing? Does it need an explanatory comment or is this copied over from Kubernetes?

Reply:
It's copied over from the Kubernetes codebase. Roughly:

* Set CODEGEN_PKG to CODEGEN_PKG if it *is* already set
* Otherwise, `cd` into the root of the repo (SCRIPT_ROOT) and attempt to list `./vendor/k8s.io/code-generator` if it exists, piping errors to /dev/null. If this command is successful, `./vendor/k8s.io/code-generator` is printed to stdout from `ls`, thus CODEGEN_PKG becomes `./vendor/k8s.io/code-generator`.
* Otherwise, return `../code-generator`. This final step is to deal with the case where our project is located within `$GOPATH/src/k8s.io` (or `k8s.io/kubernetes/staging/src/k8s.io/`). I copied this from the `sample-controller` project, hence this extra condition 😄 

The script itself will error regardless if this path is set incorrectly, so I'm not too bothered about falling back to a directory that likely doesn't exist.